core/app/models/spree/order_inventory.rb

```diff
@@ -67,13 +67,21 @@ def determine_target_shipment(quantity)
 
     def add_to_shipment(shipment, quantity)
       pending_units = []
-      if variant.should_track_inventory?
-        on_hand, back_order = shipment.stock_location.fill_status(variant, quantity)
-
-        on_hand.times { pending_units << shipment.set_up_inventory('on_hand', variant, order, line_item) }
-        back_order.times { pending_units << shipment.set_up_inventory('backordered', variant, order, line_item) }
+      if shipment
+        if variant.should_track_inventory?
+          on_hand, back_order = shipment.stock_location.fill_status(variant, quantity)
+
+          on_hand.times { pending_units << shipment.set_up_inventory('on_hand', variant, order, line_item) }
+          back_order.times { pending_units << shipment.set_up_inventory('backordered', variant, order, line_item) }
+        else
+          quantity.times { pending_units << shipment.set_up_inventory('on_hand', variant, order, line_item) }
+        end
       else
-        quantity.times { pending_units << shipment.set_up_inventory('on_hand', variant, order, line_item) }
+        order.create_missing_shipments
```

Comment:
What about creating the new shipments needed before calling this method, and passing the new shipment to it? I'm also a bit concerned about creating the new shipment with another logic only if the shipment passed is `nil`, which could seem unrelated. I mean that to understand why the new shipment is populated differently you need to check code somewhere else and maybe this can be improved?

Reply:
I surely can move the code that creates shipments and selects the relevant one in `#determine_target_shipment`, where, given the method name, seems a more appropriate place for doing it. The method `#add_to_shipment` can then be reverted to the original implementation. This is how the new code should look like: 

```ruby
    def determine_target_shipment(quantity)
      potential_shipments = order.shipments.select(&:ready_or_pending?)

      potential_shipments.detect do |shipment|
        shipment.include?(variant)
      end || potential_shipments.detect do |shipment|
        stock_item = shipment.stock_location.stock_item(variant.id)
        if stock_item
          stock_item.backorderable? || stock_item.count_on_hand >= quantity
        end
      end || potential_shipments.detect do |shipment|
        variant.stock_location_ids.include?(shipment.stock_location_id)
      end || begin
        order.create_missing_shipments
        order.shipments.detect do |order_shipment|
          order_shipment.line_items.include?(line_item)
        end
      end
    end
``` 
@kennyadsl would these changes address your concerns? 

Also, I'd like to write a few more tests in order to cover more different scenarios.

More in general, I think what I'm doing here is not the best possible solution, as having a method that has the side effect of _creating_ new shipments to fulfill its goal makes me frown, but I can't figure out a more straightforward solution unless we invest time in some substantial refactoring effort that may include the whole process of adding an item from the backend.

Reply:
@kennyadsl I eventually changed most of the problematic parts of the previous implementation. 

I think the new solution is much better (thank you for your suggestions!) as it creates only shipments when strictly necessary and only for the relevant line item.